File: scripts/plotIMU.py

import rosbag
import matplotlib.pyplot as plt
from sensor_msgs.msg import Imu

# Function to read IMU data from a rosbag file
def read_imu_data(bag_file):
    accel_data = {'x': [], 'y': [], 'z': [], 'time': []}
    ang_vel_data = {'x': [], 'y': [], 'z': [], 'time': []}

    bag = rosbag.Bag(bag_file)
    for topic, msg, t in bag.read_messages(topics=['/imu/data']):
        if isinstance(msg, Imu):
            accel_data['x'].append(msg.linear_acceleration.x)
            accel_data['y'].append(msg.linear_acceleration.y)
            accel_data['z'].append(msg.linear_acceleration.z)
            accel_data['time'].append(t.to_sec())

            ang_vel_data['x'].append(msg.angular_velocity.x)
            ang_vel_data['y'].append(msg.angular_velocity.y)
            ang_vel_data['z'].append(msg.angular_velocity.z)
            ang_vel_data['time'].append(t.to_sec())
    
    bag.close()
    return accel_data, ang_vel_data

# Function to plot IMU data
def plot_imu_data(accel_data, ang_vel_data):
    fig, axs = plt.subplots(2, 1, figsize=(12, 8))

    # Plot linear acceleration
    axs[0].plot(accel_data['time'], accel_data['x'], 'r', label='X-axis')
    axs[0].plot(accel_data['time'], accel_data['y'], 'g', label='Y-axis')
    axs[0].plot(accel_data['time'], accel_data['z'], 'b', label='Z-axis')
    axs[0].set_title('Linear Acceleration')
    axs[0].set_xlabel('Time [s]')
    axs[0].set_ylabel('Acceleration [m/s^2]')
    axs[0].legend()

    # Plot angular velocity
    axs[1].plot(ang_vel_data['time'], ang_vel_data['x'], 'r', label='X-axis')
    axs[1].plot(ang_vel_data['time'], ang_vel_data['y'], 'g', label='Y-axis')
    axs[1].plot(ang_vel_data['time'], ang_vel_data['z'], 'b', label='Z-axis')
    axs[1].set_title('Angular Velocity')
    axs[1].set_xlabel('Time [s]')
    axs[1].set_ylabel('Angular Velocity [rad/s]')
    axs[1].legend()

    plt.tight_layout()
    plt.show()

# Main function
if __name__ == "__main__":
    bag_file = '/home/larry/data/hongjing_step/test1.bag'
    accel_data, ang_vel_data = read_imu_data(bag_file)
    plot_imu_data(accel_data, ang_vel_data)

File: scripts/plotIMU.m
% clc;clear;close all;

% Specify the path to your ROS bag file
bagFilePath = '/home/larry/data/floor3_scout/2024-06-06-10-24-26.bag';
% bagFilePath = '/home/larry/data/floor3/10-290-line.bag';

% Load the ROS bag
bag = rosbag(bagFilePath);

% Select messages of type sensor_msgs/Imu
imuSelect = select(bag, 'Topic', '/ouster/imu');

% Read the IMU messages
imuMsgs = readMessages(imuSelect, 'DataFormat', 'struct');

% Initialize arrays to hold the data
timeStamps = zeros(length(imuMsgs), 1);
accelX = zeros(length(imuMsgs), 1);
accelY = zeros(length(imuMsgs), 1);
accelZ = zeros(length(imuMsgs), 1);
angVelX = zeros(length(imuMsgs), 1);
angVelY = zeros(length(imuMsgs), 1);
angVelZ = zeros(length(imuMsgs), 1);

% Extract the data from the messages
for i = 1:length(imuMsgs)
    timeStamps(i) = double(imuMsgs{i}.Header.Stamp.Sec) + double(imuMsgs{i}.Header.Stamp.Nsec) * 1e-9;
    accelX(i) = imuMsgs{i}.LinearAcceleration.X;
    accelY(i) = imuMsgs{i}.LinearAcceleration.Y;
    accelZ(i) = imuMsgs{i}.LinearAcceleration.Z;
    angVelX(i) = imuMsgs{i}.AngularVelocity.X;
    angVelY(i) = imuMsgs{i}.AngularVelocity.Y;
    angVelZ(i) = imuMsgs{i}.AngularVelocity.Z;
end

% Convert timestamps to relative time (starting from zero)
timeStamps = timeStamps - timeStamps(1);

% Plot the IMU data
figure;

% Plot linear acceleration
subplot(2, 1, 1);
plot(timeStamps, accelX, 'r', 'DisplayName', 'X-axis');
hold on;
plot(timeStamps, accelY, 'g', 'DisplayName', 'Y-axis');
plot(timeStamps, accelZ, 'b', 'DisplayName', 'Z-axis');
hold off;
title('Linear Acceleration');
xlabel('Time [s]');
ylabel('Acceleration [m/s^2]');
legend;

% Plot angular velocity
subplot(2, 1, 2);
plot(timeStamps, angVelX, 'r', 'DisplayName', 'X-axis');
hold on;
plot(timeStamps, angVelY, 'g', 'DisplayName', 'Y-axis');
plot(timeStamps, angVelZ, 'b', 'DisplayName', 'Z-axis');
hold off;
title('Angular Velocity');
xlabel('Time [s]');
ylabel('Angular Velocity [rad/s]');
legend;
